public/app/features/dashboard/utils/dashboard.ts

```diff
@@ -0,0 +1,91 @@
+import { chain, cloneDeep, defaults, find } from 'lodash';
+
+import { PanelPluginMeta } from '@grafana/data';
+import config from 'app/core/config';
+import { LS_PANEL_COPY_KEY } from 'app/core/constants';
+import store from 'app/core/store';
+import { DashboardModel, PanelModel } from 'app/features/dashboard/state';
+import { calculateNewPanelGridPos } from 'app/features/dashboard/utils/panel';
+
+export const onCreateNewPanel = (dashboard: DashboardModel): number | undefined => {
+  const newPanel: Partial<PanelModel> = {
+    type: 'timeseries',
+    title: 'Panel Title',
+    gridPos: calculateNewPanelGridPos(dashboard),
+  };
+
+  dashboard.addPanel(newPanel);
+  return newPanel.id;
+};
+
+export const onCreateNewRow = (dashboard: DashboardModel) => {
+  const newRow = {
+    type: 'row',
+    title: 'Row title',
+    gridPos: { x: 0, y: 0 },
+  };
+
+  dashboard.addPanel(newRow);
+};
+
+export const onAddLibraryPanel = (dashboard: DashboardModel) => {
+  const newPanel = {
+    type: 'add-library-panel',
+    gridPos: calculateNewPanelGridPos(dashboard),
+  };
+
+  dashboard.addPanel(newPanel);
+};
+
+type PanelPluginInfo = { defaults: { gridPos: { w: number; h: number }; title: string } };
```

Comment:
A very similar type is defined in AddPanelWidget. Can we re-use that one?

Reply:
the `id` of `PanelPluginInfo` in `AddPanelWidget` is of type `number`, whereas we are [looking for a panel in panels here](https://github.com/grafana/grafana/blob/polibb/dashboard-add-btn/public/app/features/dashboard/components/AddPanelWidget/AddPanelWidget.tsx#L47) which are of type `PanelPluginMeta` which has the `id` of type `string`. (I suppose that's why `const pluginInfo` is set of type `any`, but that's not possible anymore - _betterer_ is failing if I recreate the same method.)
The `id` is used as a holder for the panel's `type` in this search for the copied panel and then later for the [creation of the pasted panel](https://github.com/grafana/grafana/blob/polibb/dashboard-add-btn/public/app/features/dashboard/components/AddPanelWidget/AddPanelWidget.tsx#L92), so it suits it to be a string 😄.

I _think_ we'll delete this `AddPanelWidget` alltogether when the feature flag `emptyDashboardPage` is removed?


Tbh, as a whole I was confused by this method - it's creating an array, but only ever fills it with 1 value; I'm not sure if `pluginCopy.sort = -1;` is necessary and why. The main reason I think `any` was used is because the `defaults` key is set onto the `panelCopy` (`pluginCopy.defaults = copiedPanel;`), but overall I think we can either return a tuple or this kind of an object of combined types which are just the panel type and a type that cares about `defaults`.